File: native/android/app/src/main/java/com/yijing/instrument/ExperienceApi.kt
package com.yijing.instrument

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

class ExperienceApi(
    private val baseUrl: String = BuildConfig.EXPERIENCE_API_BASE_URL.trimEnd('/'),
) {
    suspend fun createSession(question: String): CreatedSession = withContext(Dispatchers.IO) {
        val json = request(
            path = "/v1/cast-sessions",
            method = "POST",
            body = JSONObject().put("question", question),
        )
        CreatedSession(json.toSessionSnapshot(), json.getString("session_token"))
    }

    suspend fun getSession(id: String, token: String): SessionSnapshot = withContext(Dispatchers.IO) {
        request("/v1/cast-sessions/$id", token = token).toSessionSnapshot()
    }

    suspend fun lockNextLine(
        id: String,
        token: String,
        idempotencyKey: String,
        energy: Double,
        triggerMode: String,
    ): LockedLineResponse = withContext(Dispatchers.IO) {
        val json = request(
            path = "/v1/cast-sessions/$id/lines",
            method = "POST",
            token = token,
            headers = mapOf("Idempotency-Key" to idempotencyKey),
            body = JSONObject()
                .put("motion_energy", energy.coerceIn(0.0, 100.0))
                .put("trigger_mode", triggerMode),
        )
        LockedLineResponse(
            status = json.getString("status"),
            lineCount = json.getInt("line_count"),
            line = CastLine.fromJson(json.getJSONObject("line")),
        )
    }

    suspend fun complete(id: String, token: String): CompletionResponse = withContext(Dispatchers.IO) {
        request("/v1/cast-sessions/$id/complete", method = "POST", token = token).toCompletion()
    }

    suspend fun result(id: String, token: String): CompletionResponse = withContext(Dispatchers.IO) {
        request("/v1/cast-sessions/$id/result", token = token).toCompletion()
    }

    private fun request(
        path: String,
        method: String = "GET",
        token: String? = null,
        headers: Map<String, String> = emptyMap(),
        body: JSONObject? = null,
    ): JSONObject {
        val connection = (URL("$baseUrl$path").openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 12_000
            readTimeout = 25_000
            setRequestProperty("Accept", "application/json")
            token?.let { setRequestProperty("Authorization", "Bearer $it") }
            headers.forEach(::setRequestProperty)
            if (body != null) {
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
                outputStream.bufferedWriter(Charsets.UTF_8).use { it.write(body.toString()) }
            }
        }
        try {
            val status = connection.responseCode
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            val text = stream?.bufferedReader(Charsets.UTF_8)?.use { it.readText() }.orEmpty()
            if (status !in 200..299) {
                val message = runCatching { JSONObject(text).optString("detail") }.getOrNull()
                    .takeUnless { it.isNullOrBlank() }
                    ?: "服务暂时不可用（$status）。"
                throw ExperienceApiException(status, message)
            }
            return JSONObject(text)
        } finally {
            connection.disconnect()
        }
    }
}

class ExperienceApiException(val status: Int, message: String) : Exception(message)

File: native/ios/DivinationApp/Services/ExperienceAPI.swift
import Foundation

enum ExperienceAPIError: LocalizedError {
    case invalidResponse
    case server(Int, String)

    var errorDescription: String? {
        switch self {
        case .invalidResponse:
            return "服务响应格式异常，请稍后重试。"
        case let .server(_, message):
            return message
        }
    }
}

actor ExperienceAPI {
    static let shared = ExperienceAPI()

    private let baseURL: URL
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    init() {
        let configured = ProcessInfo.processInfo.environment["EXPERIENCE_API_BASE_URL"]
        baseURL = URL(string: configured ?? "https://ai-divination-experience.onrender.com")!
    }

    func createSession(question: String) async throws -> CreatedSession {
        try await request(
            path: "/v1/cast-sessions",
            method: "POST",
            body: ["question": question]
        )
    }

    func getSession(id: String, token: String) async throws -> SessionSnapshot {
        try await request(path: "/v1/cast-sessions/\(id)", token: token)
    }

    func lockNextLine(
        id: String,
        token: String,
        idempotencyKey: String,
        energy: Double,
        triggerMode: String
    ) async throws -> LockedLineResponse {
        try await request(
            path: "/v1/cast-sessions/\(id)/lines",
            method: "POST",
            token: token,
            headers: ["Idempotency-Key": idempotencyKey],
            body: ["motion_energy": energy, "trigger_mode": triggerMode]
        )
    }

    func complete(id: String, token: String) async throws -> CompletionResponse {
        try await request(
            path: "/v1/cast-sessions/\(id)/complete",
            method: "POST",
            token: token
        )
    }

    func result(id: String, token: String) async throws -> CompletionResponse {
        try await request(path: "/v1/cast-sessions/\(id)/result", token: token)
    }

    private func request<Response: Decodable>(
        path: String,
        method: String = "GET",
        token: String? = nil,
        headers: [String: String] = [:],
        body: [String: Any]? = nil
    ) async throws -> Response {
        var request = URLRequest(url: baseURL.appending(path: path))
        request.httpMethod = method
        request.timeoutInterval = 25
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        for (name, value) in headers {
            request.setValue(value, forHTTPHeaderField: name)
        }
        if let body {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ExperienceAPIError.invalidResponse
        }
        guard 200..<300 ~= http.statusCode else {
            let payload = try? decoder.decode(APIErrorPayload.self, from: data)
            throw ExperienceAPIError.server(
                http.statusCode,
                payload?.detail ?? "服务暂时不可用（\(http.statusCode)）。"
            )
        }
        do {
            return try decoder.decode(Response.self, from: data)
        } catch {
            throw ExperienceAPIError.invalidResponse
        }
    }
}

private struct APIErrorPayload: Decodable {
    let detail: String
}
